File: py/candlestick_patterns/patterns/marubozu.py
"""Marubozu pattern (1-candle)."""
from __future__ import annotations

from ..core.primitives import is_white, real_body, upper_shadow, lower_shadow


def marubozu(self) -> int:
    """Marubozu: a one-candle pattern.

    Must have:
    - long real body,
    - very short upper shadow,
    - very short lower shadow.

    The meaning of "long" is specified with self._long_body.
    The meaning of "very short" for shadows is specified with
    self._very_short_shadow.

    Returns:
        +100 for bullish (white), -100 for bearish (black), 0 for no pattern.
    """
    if not self._enough(1, self._long_body, self._very_short_shadow):
        return 0

    o, h, l, c = self._bar(1)

    vs = self._avg(self._very_short_shadow, 1)

    if (real_body(o, c) > self._avg(self._long_body, 1) and
            upper_shadow(o, h, c) < vs and
            lower_shadow(o, l, c) < vs):
        return 100 if is_white(o, c) else -100

    return 0

File: go/candlestickpatterns/patterns/marubozu.go
package patterns

import (
	"zpano/candlestickpatterns/core"
	"zpano/fuzzy"
)

// Marubozu detects the Marubozu pattern (1-candle).
//
// Must have:
//   - long real body,
//   - very short upper shadow,
//   - very short lower shadow.
//
// The meaning of "long" is specified with cp.LongBody.
// The meaning of "very short" for shadows is specified with cp.VeryShortShadow.
//
// Category B: direction from candle color.
func Marubozu(cp *core.CandlestickPatterns) float64 {
	if !cp.Enough(1, cp.LongBody, cp.VeryShortShadow) {
		return 0.0
	}

	b := cp.Bar(1)
	// Fuzzy: long body, very short shadows.
	muLong := cp.MuGreater(core.RealBodyLen(b.O, b.C), cp.LongBody, 1)
	muUS := cp.MuLess(core.UpperShadow(b.O, b.H, b.C), cp.VeryShortShadow, 1)
	muLS := cp.MuLess(core.LowerShadow(b.O, b.L, b.C), cp.VeryShortShadow, 1)

	confidence := fuzzy.TProductAll(muLong, muUS, muLS)
	// Crisp direction from color.
	direction := 1
	if !core.IsWhite(b.O, b.C) {
		direction = -1
	}
	return float64(direction) * confidence * 100.0
}
